File: Problem_116_Populating_Next_Right_Pointers_in_Each_Node/solution.py
from typing import Optional


class Node:
    def __init__(self, val: int = 0, left: Optional['Node'] = None, right: Optional['Node'] = None, next: Optional['Node'] = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


def connect(root: Optional[Node]) -> Optional[Node]:
    if not root:
        return None

    leftmost = root
    while leftmost.left:
        current = leftmost
        while current:
            current.left.next = current.right
            if current.next:
                current.right.next = current.next.left
            current = current.next
        leftmost = leftmost.left

    return root


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    root.right.left = Node(6)
    root.right.right = Node(7)
    connect(root)
    print(root.left.left.next.val)

File: Problem_116_Populating_Next_Right_Pointers_in_Each_Node/solution.cpp
#include <cstddef>
using namespace std;

class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(nullptr), right(nullptr), next(nullptr) {}
    Node(int _val) : val(_val), left(nullptr), right(nullptr), next(nullptr) {}
    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};

class Solution {
public:
    Node* connect(Node* root) {
        if (root == nullptr) {
            return nullptr;
        }

        Node* leftmost = root;
        while (leftmost->left != nullptr) {
            Node* current = leftmost;
            while (current != nullptr) {
                current->left->next = current->right;
                if (current->next != nullptr) {
                    current->right->next = current->next->left;
                }
                current = current->next;
            }
            leftmost = leftmost->left;
        }

        return root;
    }
};
